src/main/java/org/jenkinsci/plugins/docker/workflow/RegistryEndpointStep.java

```diff
@@ -105,6 +111,28 @@ public DescriptorImpl() {
             return true;
         }
 
+        @Override public UninstantiatedDescribable uninstantiate(Step step) throws UnsupportedOperationException {
+            RegistryEndpointStep s = (RegistryEndpointStep) step;
+            Map<String, Object> args = new TreeMap<>();
+            args.put("url", s.registry.getUrl());
+            args.put("credentialsId", s.registry.getCredentialsId());
+            args.put("toolName", s.toolName);
+            args.values().removeAll(Collections.singleton(null));
+            return new UninstantiatedDescribable(args);
+        }
+
+        @Override public Step newInstance(Map<String, Object> arguments) throws Exception {
+            if (arguments.containsKey("url") || arguments.containsKey("credentialsId")) {
+                if (arguments.containsKey("registry")) {
+                    throw new IllegalArgumentException("cannot mix url/credentialsId with registry");
+                }
+                arguments.put("registry", new DockerRegistryEndpoint((String) arguments.remove("url"), (String) arguments.remove("credentialsId")));
```

Comment:
Will this actually work if you have a credential but not a URL?  Or vice versa?

If not, should print something reasonable as an error.

Reply:
Yes,

```groovy
withDockerRegistry(credentialsId: '…') {…}
```

logs you in to Docker Hub rather than a private registry. Conversely, if there were some way to have a non-password-protected private registry (or if it had read-only access and you were not pushing anything), you could

```groovy
withDockerRegistry(url: '…') {…}
```